src/ucp/core/ucp_ep.c

```diff
@@ -835,27 +843,44 @@ void ucp_ep_destroy_internal(ucp_ep_h ep)
 
 static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
 {
+    ucp_ep_ext_gen_t UCS_V_UNUSED *ep_ext = ucp_ep_ext_gen(ep);
     ucp_lane_index_t lane;
     uct_ep_h uct_ep;
+#if UCS_ENABLE_ASSERT
+    unsigned num_inprog;
+#endif
 
     for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
         uct_ep        = ep->uct_eps[lane];
         uct_eps[lane] = uct_ep;
 
-        /* an attempt to destroy UCT lanes can be done several times inside
+        /* An attempt to destroy UCT lanes can be done several times inside
          * ucp_ep_cleanup_lanes()/ucp_ep_discard_lanes(), i.e. discarding lanes
          * and then cleanup lanes during UCP EP close - so, if it happens, UCT
          * lane has to be already set to '&ucp_failed_tl_ep' and UCP EP's
          * 'ref_cnt' can be >= 1, when UCP EP is going to be destroyed.
          * Otherwise - UCT lane is set to a real UCT EP and it means that
-         * discarding of UCT lanes was not started yet,  i.e. called from UCP EP
-         * destroy function and UCP EP's ref_cnt has to be '1'.
+         * discarding of UCT lanes was not started yet, i.e. called from UCP EP
+         * destroy function and UCP EP's ref_cnt has to be ('number of worker
+         * flush operations started for the UCP EP' + 'number of discarded
+         * operations started for the UCP EP' + '1').
          * If UCT lane is destroyed, but some operations are still in-progress,
-         * acompletions can be received for his UCT EP - this will lead to
+         * completions can be received for this UCT EP - this will lead to
          * undefined behavior, because a real UCT EP was already destroyed */
-        ucs_assert(ucp_is_uct_ep_failed(uct_ep) || (ep->ref_cnt == 1));
+#if UCS_ENABLE_ASSERT
+        num_inprog = ep_ext->num_discard_inprog + ep_ext->num_flush_inprog;
+        if (ucp_is_uct_ep_failed(uct_ep)) {
+            ucs_assert((ep->ref_cnt == num_inprog) ||
+                       (ep->ref_cnt == (num_inprog + 1)));
+        } else {
+            ucs_assert(ep->ref_cnt == (num_inprog + 1));
+        }
+#endif
+        
+        ucs_assert(ucp_is_uct_ep_failed(uct_ep) ||
+                   (ep->ref_cnt == (ucp_ep_ext_gen(ep)->num_flush_inprog + 1)));
```

Comment:
seems like duplication of lines 870-878. Also can use `ep_ext
`

Reply:
fixed